lib/web3-contracts.js

```diff
@@ -131,248 +106,189 @@ export function useTokenBalance(symbol) {
+              antTreasuryBalance,
+              CONNECTOR_WEIGHT,
+              amount
+            )
+          : bancorContract.calculateSaleReturn(
+              totalSupply,
+              antTreasuryBalance,
+              CONNECTOR_WEIGHT,
+              amount
+            ))
+        if (!cancelled) {
+          setLoading(false)
+          setPrice(salePrice)
+        }
+      } catch (err) {
+        if (!cancelled) {
+          retryTimer = setTimeout(useBondingCurvePrice, RETRY_EVERY)
+        }
       }
     }
-    wrapperContract.on('Bought', onBought)
+
+    getSalePrice()
 
     return () => {
-      wrapperContract.removeListener('Bought', onBought)
+      cancelled = true
+      clearTimeout(retryTimer)
     }
-  }, [account, wrapperContract, updateBalance])
-
-  return balance
+  }, [
+    amount,
+    antTreasuryBalance,
+    bancorContract,
+    forwards,
+    supplyLoading,
+    totalSupply,
+  ])
+
+  return useMemo(() => ({ loading, price }), [loading, price])
 }
 
 // Convert ANT to ANJ action
-export function useConvertAntToAnj() {
+export function useOpenOrder() {
+  const { account } = useWeb3Connect()
   const antContract = useKnownContract('TOKEN_ANT')
-  const [wrapperAddress] = getKnownContract('WRAPPER')
+  const fundraisingContract = useKnownContract('FUNDRAISING')
+
+  const [antAddress] = getKnownContract('TOKEN_ANT')
+  const [marketMakerAddress] = getKnownContract('MARKET_MAKER')
 
   return useCallback(
-    async antAmount => {
-      if (!antContract || !wrapperAddress) {
+    async (amount, forwards = true) => {
+      if (!antContract || !fundraisingContract) {
         return false
       }
+      // Interacting with the bonding curve involves 3 transactions (or two,
+      // depending on the direction):
+      // 1. Approval (if there's not enough allowance and we're converting ANT -> ANJ)
+      // 2. Open a buy order (and wait for the transaction to be mined)
+      // 3. Claim the order (done in useClaimOrder)
+
+      // We first check for allowance if the direction is correct,
+      // and if we need more, approve
+      if (forwards) {
+        const allowance = await antContract.allowance(
+          account,
+          marketMakerAddress
+        )
+        if (allowance.lt(bigNum(amount))) {
+          try {
+            await antContract.approve(marketMakerAddress, amount)
+            // Don't wait for the approval to be mined before showing second transaction
+          } catch (err) {
+            throw new Error('User did not approve transaction')
+          }
+        }
+      }
 
-      return antContract.approveAndCall(wrapperAddress, antAmount, '0x00', {
-        gasLimit: 1000000,
-      })
+      // Then, we open a buy or sell order. The next step will be handled after the
+      // transaction gets mined, in useClaimOrder()
+      return await (forwards
+        ? fundraisingContract.openBuyOrder(antAddress, amount, {
+            gasLimit: 850000,
+            value: 0,
```

Comment:
Interesting bit here: I haven't sanity checked gas prices (should be on our checklist for final deployment, and we should track these on tenderly), but, if we don't set the `gasLimit` & `value` overrides, this will fail, saying "Always failing transaction".

Reply:
Having to set the `value` shouldn't be needed, but if we're sending the transaction immediately before the approve is mined, the "Always failing transaction" is Metamask's normal behaviour. We only hardcode gas amounts in other places to avoid this error.

If we wait between the approve and open order, this error should not appear.